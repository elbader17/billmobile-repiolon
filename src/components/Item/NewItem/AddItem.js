import React from 'react';
import { View, TextInput, Text } from 'react-native';
import PropTypes from 'prop-types';
import style from '../style';

const AddItem = props => {
  if (props.isProduct) {
    return (
      <View>
        <View style={style.textBoxBtnHolder}>
          <TextInput
            placeholder="Nombre de Producto"
            value={props.data.nameProduct}
            onChangeText={props.setName}
            style={style.textName}
          />
        </View>
        <View style={style.price}>
          <Text style={style.textRegular14GrayDarkPrice}>Precio $ </Text>
            <TextInput
              placeholder="0.00"
              value={props.data.price}
              onChangeText={props.setValueOrPrice}
              style={style.textBoxPrice}
              keyboardType='numeric'
            />
        </View>
      </View>
    );
  } else {
    return (
      <View>
        <View style={style.textBoxBtnHolder}>
          <TextInput
            placeholder="Concepto del Servicio"
            value={props.data.nameService}
            onChangeText={props.setName}
            style={style.textName}
          />
        </View>
        <View style={style.price}>
          <Text style={style.textRegular14GrayDarkPrice}>Valor $ </Text>
          <TextInput
            placeholder="0.00"
            value={props.data.value}
            onChangeText={props.setValueOrPrice}
            style={style.textBoxPrice}
            keyboardType='numeric'
          />
        </View>
      </View>
    );
  }
}

AddItem.propTypes = {
  isProduct: PropTypes.bool,
  setName: PropTypes.func,
  setValueOrPrice: PropTypes.func,
  data: PropTypes.object
};

export default AddItem;