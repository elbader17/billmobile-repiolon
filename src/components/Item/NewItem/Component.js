import React from 'react';
import { View, Text, TouchableOpacity } from 'react-native';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
import { Button } from "react-native-elements";
import LinearGradient from 'react-native-linear-gradient';
import Icon from 'react-native-vector-icons/AntDesign';
import { GRADIANTBLUE2, COLORGB, COLORGB2 } from '../../../constants/colors';
import { validateAddItem } from '../../../utils/validations';
import AddItem from './AddItem';
import style from '../style';

class NewItem extends React.Component {
  constructor(props) {
    super(props);
    const item = this.props.navigation.getParam('item', this.defaultItem());
    const productOrService = item.attributes.category==='product';
    const isProducts = this.props.navigation.getParam('isProduct', productOrService );
    this.state = {
      name: item.attributes.name,
      price: item.attributes.price,
      isProduct: isProducts,
      itemId: item.id,
      loading: false,
      errorName: undefined, 
      errorPrice: undefined
    };
  }

  static navigationOptions = ({navigation}) => {
    return {
      title: 'Cargar Items',
      headerBackground: (
        <LinearGradient
          colors={ GRADIANTBLUE2 }
          style={{ flex: 1 }}
          start={{x: 0, y: 0}}
          end={{x: 1, y: 0}}
        />
      ),
      headerTitleStyle: style.headerText,
      headerTintColor: 'white',
      headerLeft: <TouchableOpacity onPress={()=> {
                    if (navigation.state.params.type === 'collection') navigation.navigate('ItemList');
                    else navigation.navigate('ListInvoiceItem');
                  }}>
                    <Icon name="left" size={20} color="white" style={{marginLeft:20}}/>
                  </TouchableOpacity> 
    }
  };

  componentWillMount() {
    this.props.navigation.setParams({type: this.props.type}); //Use in Header Left Navigation
  }

  defaultItem = () => {
    return {
      attributes: {
        name: '',
        price: '',
        category: 'product',
      }
    };
  }

  saveItem = () => {
    const { name, price, isProduct, itemId } = this.state;
    const { saveItem, type, navigation } = this.props;
    const category = isProduct ? 'product' : 'service';
    const quantity = 1;
    if (validateAddItem(name, price)) {
      this.setState({ isEnableButton: false});
      this.setLoading(true);
      saveItem(
        {
          id: itemId,
          category,
          name,
          price,
          quantity
        },
        navigation, type
      );
    }
    else {
      if (name=='' && price=='') 
        this.setState({
          errorPrice: 'Debe ingresar un valor', 
          errorName: 'Este campo no puede ser vacio'
        })  
      else if (price=='') 
        this.setState({errorPrice: 'Debe ingresar un valor'})
        else 
          this.setState({errorName: 'Este campo no puede ser vacio'})
          
    }
  }

  setName = (value) => this.setState({ name: value })
  setPrice = (value) => this.setState({ price:value })
  setLoading = (bool) => this.setState({ loading: bool })
  setErrorName = () => this.setState({errorName: undefined})
  setErrorPrice = () => this.setState({errorPrice: undefined})

  render() {
    const data = {
      isProduct: this.state.isProduct,
      name: this.state.name,
      price: this.state.price,
      errorName: this.state.errorName,
      errorPrice: this.state.errorPrice
    }
    return(
      <KeyboardAwareScrollView>
        <View style={style.container}>

          <View style={style.containerBody}>
            <View style={[style.boxSelectButton, style.inLineSpaceAround]}>
              <Button
                title='Productos'
                onPress={() => this.setState({isProduct: true}) }
                buttonStyle={ style.buttonSelect }
                titleStyle={ style.textRegular16White }
                ViewComponent={LinearGradient}
                linearGradientProps={this.state.isProduct ? COLORGB : COLORGB2}
              />
              <Button
                title='Servicios'
                onPress={() => this.setState({isProduct: false}) }
                buttonStyle={ style.buttonSelect }
                titleStyle={ style.textRegular16White }
                ViewComponent={LinearGradient}
                linearGradientProps={this.state.isProduct ? COLORGB2 : COLORGB}
              />
            </View>

            <View style={style.boxInput}>
              <AddItem
                data={data}
                setName={this.setName}
                setPrice={this.setPrice}
                setErrorName={this.setErrorName}
                setErrorPrice={this.setErrorPrice}
              />
            </View>
          </View> 

          <View style={[style.containerFooter, {alignItems: 'center'}]}>
            <Button
              title='Guardar'
              onPress={ this.saveItem }
              buttonStyle={ style.buttonSave }
              titleStyle={ style.textRegular16White }
              loading={this.state.loading}
              ViewComponent={LinearGradient}
              linearGradientProps={COLORGB2}
            /> 
          </View> 

        </View>
      </KeyboardAwareScrollView>
    );
  }
}

export default NewItem;