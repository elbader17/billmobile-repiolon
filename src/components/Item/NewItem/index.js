import { connect } from 'react-redux';
import NewItem from './Component';
import { registerItemProduct, registerItemService } from '../../../app/items/action';

const mapStateToProps = state => ({
  jwtToken: state.authentication.jwtToken,
  name: state.addItem.name,
});

function mapDispatchToProps(dispatch) {
  return {
    registerItemProduct: (name, price, jwtToken) => dispatch(registerItemProduct(name, price, jwtToken)),
    registerItemService: (name, price, unit, jwtToken) => dispatch(registerItemService(name, price, unit, jwtToken)),
  };
}

const component = connect(
  mapStateToProps,
  mapDispatchToProps,
)(NewItem);

export default component;
