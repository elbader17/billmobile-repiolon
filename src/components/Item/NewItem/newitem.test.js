import React from 'react';
import { shallow } from 'enzyme';
import AddItem from './AddItem';
import NewItem from './Component';

jest.mock('../../../constants/metrics', () => ({ METRICS: { heightHeader: {} } }));

const promise = Promise.resolve(0);
const createItemMock = jest.fn(() => promise);

const setupAddItem = () => {
  const props = {
    data: {
      nameProduct: '',
      price: '',
      nameService:'',
      value:''
    } 
  }
  const functionalComponent = shallow(<AddItem {...props}/>);
  return {props, functionalComponent};
};

describe('test functional component AddItem', () => {
  test('render correctly component AddItem', () => {
    const {functionalComponent} = setupAddItem();
    expect(functionalComponent).toMatchSnapshot();
  });
});

test('render correctly component NewItem', () => {
  const component = shallow(<NewItem />)
  expect(component).toMatchSnapshot();
});

test('call the createItem function with param product', () => {
  const component = shallow(<NewItem createItem={createItemMock}/>);
  const button = component.findWhere(node => node.prop('testID') === 'buttonSave');
  button.simulate('press');
  expect(createItemMock).toBeCalled();
});

test('call the createItem function with param service', () => {
  const component = shallow(<NewItem createItem={createItemMock}/>);
  component.setState({isProduct: false});
  const button = component.findWhere(node => node.prop('testID') === 'buttonSave');
  button.simulate('press');
  expect(createItemMock).toBeCalled();
});

test('initial state of isProduct is true', () => {
  const component = shallow(<NewItem />);
  expect(component.instance().state.isProduct).toBe(true);
});

describe('check title button save item', () => {
  test('state isProduct true - button Text GUARDAR PRODUCTO', () => {
    const component = shallow(<NewItem />);
    const button = component.findWhere(node => node.prop('testID') === 'buttonSave');
    const title = [ 'GUARDAR ', 'PRODUCTO' ];
    expect(button.props().title.props.children).toEqual(title);
  });

  test('state isProduct false - button Text GUARDAR SERVICIO', () => {
    const component = shallow(<NewItem />);
    component.setState({isProduct: false});
    const button = component.findWhere(node => node.prop('testID') === 'buttonSave');
    const title = [ 'GUARDAR ', 'SERVICIO' ];
    expect(button.props().title.props.children).toEqual(title);
  });
});

describe('check correctly setState button save', () => {
  test('product button sets the state isProduct to true', () => {
    const component = shallow(<NewItem />);
    component.setState({isProduct: false});
    const button = component.findWhere(node => node.prop('testID') === 'product');
    button.simulate('press');
    expect(component.instance().state.isProduct).toBe(true);
  });

  test('service button sets the state isProduct to false', () => {
    const component = shallow(<NewItem />);
    const button = component.findWhere(node => node.prop('testID') === 'service');
    button.simulate('press');
    expect(component.instance().state.isProduct).toBe(false);
  });
});

describe('check disabled button save', () => {
  test('button disabled at startup', () => {
    const component = shallow(<NewItem />);
    const button = component.findWhere(node => node.prop('testID') === 'buttonSave');
    expect(button.props().disabled).toBe(true);
  });

  test('GUARDAR PRODUCTO button enabled with the correct data', () => {
    const component = shallow(<NewItem />);
    component.setState({nameProduct:'Coca Cola', priceProduct:'123', isProduct: true,});
    const button = component.findWhere(node => node.prop('testID') === 'buttonSave');
    expect(button.props().disabled).toBe(false);
  });

  test('GUARDAR PRODUCTO button disabled with nameProduct empty', () => {
    const component = shallow(<NewItem />);
    component.setState({nameProduct:'', priceProduct:'123', isProduct: true,});
    const button = component.findWhere(node => node.prop('testID') === 'buttonSave');
    expect(button.props().disabled).toBe(true);
  });

  test('GUARDAR PRODUCTO button disabled with price empty', () => {
    const component = shallow(<NewItem />);
    component.setState({nameProduct:'Coca Cola', priceProduct:'', isProduct: true,});
    const button = component.findWhere(node => node.prop('testID') === 'buttonSave');
    expect(button.props().disabled).toBe(true);
  });

  test('GUARDAR SERVICIO button enabled with the correct data', () => {
    const component = shallow(<NewItem />);
    component.setState({nameService:'Flete', valueService:'123', isProduct: false,});
    const button = component.findWhere(node => node.prop('testID') === 'buttonSave');
    expect(button.props().disabled).toBe(false);
  });

  test('GUARDAR SERVICIO button disabled with nameService empty', () => {
    const component = shallow(<NewItem />);
    component.setState({nameService:'', valueService:'123', isProduct: false,});
    const button = component.findWhere(node => node.prop('testID') === 'buttonSave');
    expect(button.props().disabled).toBe(true);
  });

  test('GUARDAR SERVICIO button disabled with value empty', () => {
    const component = shallow(<NewItem />);
    component.setState({nameService:'Flete', valueService:'', isProduct: false,});
    const button = component.findWhere(node => node.prop('testID') === 'buttonSave');
    expect(button.props().disabled).toBe(true);
  });
});