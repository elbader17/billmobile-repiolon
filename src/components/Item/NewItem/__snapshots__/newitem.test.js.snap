// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly component NewItem 1`] = `
<KeyboardAvoidingView
  behavior="padding"
  enabled={true}
  keyboardVerticalOffset={Object {}}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ScrollViewMock>
    <View
      style={
        Object {
          "backgroundColor": "white",
          "height": NaN,
          "paddingHorizontal": 18,
          "paddingVertical": 10,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "marginHorizontal": 20,
              "marginVertical": 12,
            },
            Object {
              "flexDirection": "row",
              "justifyContent": "space-around",
            },
          ]
        }
      >
        <ForwardRef(Themed.Button)
          buttonStyle={
            Object {
              "backgroundColor": "#EE6123",
              "borderRadius": 5,
              "height": 66.5,
              "width": 267,
            }
          }
          onPress={[Function]}
          testID="product"
          title="PRODUCTO"
          titleStyle={
            Object {
              "color": "#fff",
              "fontFamily": "Lato-Semibold",
              "fontSize": 12,
            }
          }
        />
        <ForwardRef(Themed.Button)
          buttonStyle={
            Object {
              "backgroundColor": "transparent",
              "borderColor": "#3687d1",
              "borderRadius": 5,
              "borderWidth": 1,
              "height": 66.5,
              "width": 267,
            }
          }
          onPress={[Function]}
          testID="service"
          title="SERVICIO"
          titleStyle={
            Object {
              "color": "#3687d1",
              "fontFamily": "Lato-Semibold",
              "fontSize": 12,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "backgroundColor": "transparent",
            "marginHorizontal": 10,
          }
        }
      >
        <AddItem
          data={
            Object {
              "nameProduct": "",
              "nameService": "",
              "price": "",
              "value": "",
            }
          }
          isProduct={true}
          setName={[Function]}
          setValueOrPrice={[Function]}
        />
      </View>
    </View>
  </ScrollViewMock>
  <ForwardRef(Themed.Button)
    buttonStyle={
      Object {
        "backgroundColor": "#EE6123",
        "borderRadius": 0,
        "height": 106.5,
      }
    }
    disabled={true}
    disabledStyle={
      Object {
        "backgroundColor": "#C4C4C4",
        "borderRadius": 0,
        "height": 106.5,
      }
    }
    disabledTitleStyle={
      Object {
        "color": "#fff",
        "fontFamily": "Lato-Semibold",
        "fontSize": 14,
      }
    }
    loading={false}
    onPress={[Function]}
    testID="buttonSave"
    title={
      <Text>
        GUARDAR 
        PRODUCTO
      </Text>
    }
    titleStyle={
      Object {
        "color": "#fff",
        "fontFamily": "Lato-Semibold",
        "fontSize": 14,
      }
    }
  />
</KeyboardAvoidingView>
`;

exports[`test functional component AddItem render correctly component AddItem 1`] = `
<View>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "justifyContent": "center",
        "marginBottom": 10,
        "marginTop": 5,
        "position": "relative",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      placeholder="Concepto del Servicio"
      rejectResponderTermination={true}
      style={
        Object {
          "alignSelf": "stretch",
          "borderBottomColor": "#C4C4C4",
          "borderBottomWidth": 1,
          "borderRadius": 5,
          "fontFamily": "Lato-Regular",
          "fontSize": 18,
          "height": 40,
        }
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderColor": "#C4C4C4",
        "borderRadius": 3,
        "borderWidth": 1,
        "flexDirection": "row",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#858585",
          "fontFamily": "Lato-Regular",
          "fontSize": 18,
          "marginHorizontal": 7,
          "paddingBottom": 3,
        }
      }
    >
      Precio $ 
    </Text>
    <TextInput
      allowFontScaling={true}
      keyboardType="numeric"
      placeholder="0.00"
      rejectResponderTermination={true}
      style={
        Object {
          "fontFamily": "Lato-Regular",
          "fontSize": 18,
          "height": 93.5,
          "paddingRight": 7,
          "width": "80%",
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
</View>
`;
